[skwasm] Port to `DisplayList` objects

eyebrowsoffire opened the pull request:
This PR refactors the skwasm renderer to use `DisplayList` objects as its main model objects instead of using Skia objects directly. Then, at render time, we dispatch the display list commands to the skia surface. This is a preparatory step for impeller on web.
* Some build rules were reworked in order to allow `DisplayList` to compile via emscripten
* Some pieces of the display list library were further refactored to allow us to compile it without actually building and linking the impeller shaders. The two major classes that needed to be separated out were `DlRuntimeEffect` and the text drawing system.
* `SkPath` and `SkImage` are still used as the main model objects in skwasm. As of right now, `DisplayList` just thinly wraps these objects, so this is the minimal possible change for now. I will have to refactor this somewhat further when preparing for actual impeller adoption.
* Several special cased code paths in skwasm were removed, as they are taken care of by `DisplayList` itself. This includes shadow drawing, determining when to enable dithering, and determining the right clamp value for filters.

harryterkelsen reviewed (approved):
The `web_ui` parts LGTM. I don't know enough about the native engine to comment on the display_list changes except to note that some of them seem orthogonal to the `web_ui` changes. I think the `drawText` and runtime effects changes could be a separate PR.

eyebrowsoffire commented:
> The `web_ui` parts LGTM. I don't know enough about the native engine to comment on the display_list changes except to note that some of them seem orthogonal to the `web_ui` changes. I think the `drawText` and runtime effects changes could be a separate PR.

The `drawText` and runtime effect changes are there to decouple some impeller things from the main display list library. If I didn't do them in this change, depending on the `display_list` library would pull in a bunch of impeller stuff that doesn't yet compile on emscripten. I'd have to do a bunch more massaging to get that code working on the web build, and I'd rather do that in a subsequent PR when I'm actually adding  the impeller backend.

flar commented:
> > The `web_ui` parts LGTM. I don't know enough about the native engine to comment on the display_list changes except to note that some of them seem orthogonal to the `web_ui` changes. I think the `drawText` and runtime effects changes could be a separate PR.
> 
> The `drawText` and runtime effect changes are there to decouple some impeller things from the main display list library. If I didn't do them in this change, depending on the `display_list` library would pull in a bunch of impeller stuff that doesn't yet compile on emscripten. I'd have to do a bunch more massaging to get that code working on the web build, and I'd rather do that in a subsequent PR when I'm actually adding the impeller backend.

I thought the same thing, but you are looking at it backwards. If you land DlText first then it works, but you'd have to hold off on the skwasm changes until that PR lands. Your call, but they can be separated...

eyebrowsoffire commented:
> > > The `web_ui` parts LGTM. I don't know enough about the native engine to comment on the display_list changes except to note that some of them seem orthogonal to the `web_ui` changes. I think the `drawText` and runtime effects changes could be a separate PR.
> > 
> > 
> > The `drawText` and runtime effect changes are there to decouple some impeller things from the main display list library. If I didn't do them in this change, depending on the `display_list` library would pull in a bunch of impeller stuff that doesn't yet compile on emscripten. I'd have to do a bunch more massaging to get that code working on the web build, and I'd rather do that in a subsequent PR when I'm actually adding the impeller backend.
> 
> I thought the same thing, but you are looking at it backwards. If you land DlText first then it works, but you'd have to hold off on the skwasm changes until that PR lands. Your call, but they can be separated...

Actually yes, you're right. If you think it's important, I can try to split this PR out.

harryterkelsen commented:
> > > > > The `web_ui` parts LGTM. I don't know enough about the native engine to comment on the display_list changes except to note that some of them seem orthogonal to the `web_ui` changes. I think the `drawText` and runtime effects changes could be a separate PR.
> > > > 
> > > > 
> > > > The `drawText` and runtime effect changes are there to decouple some impeller things from the main display list library. If I didn't do them in this change, depending on the `display_list` library would pull in a bunch of impeller stuff that doesn't yet compile on emscripten. I'd have to do a bunch more massaging to get that code working on the web build, and I'd rather do that in a subsequent PR when I'm actually adding the impeller backend.
> > > 
> > > 
> > > I thought the same thing, but you are looking at it backwards. If you land DlText first then it works, but you'd have to hold off on the skwasm changes until that PR lands. Your call, but they can be separated...
> > 
> > 
> > Actually yes, you're right. If you think it's important, I can try to split this PR out.
> 
> I think it was a missed opportunity, but given the logistics of having them baked independently when they are both ready to go right now is so constrained that there isn't much point of doing it at this stage.

Agreed

flar reviewed (changes requested):
I think I reviewed all of the DL code here. Lots of suggestions.

eyebrowsoffire commented:
Since this is a big PR, and there's a nonzero chance we'll need to revert if something goes wrong, I think it actually is worthwhile to separate the display list stuff out, so I did so in this PR: https://github.com/flutter/flutter/pull/174219

I'll address Jim's comments over in that one, and then once we have that one landed I'll rebase this PR and the diffs should be smaller.

eyebrowsoffire commented on engine/src/flutter/display_list/dl_op_records.h:
```diff
@@ -1001,35 +1004,24 @@ struct DrawDisplayListOp final : DrawOpBase {
-    receiver.drawTextFrame(text_frame, x, y);
+  DisplayListCompare equals(const DrawTextOp* other) const {
+    return text->getTextBlob() == other->text->getTextBlob() &&
+                   text->getTextFrame() == other->text->getTextFrame() &&
```
I implemented this because some unit tests were failing without it. I guess the question is, should two different `DlText` objects which point to the same text blob or text frame compare as equal? If the answer is yes, I think we do need to override this. If the answer is no, I probably need to change some of the unit tests to reflect that.

eyebrowsoffire committed:
[skwasm] Port to DisplayList objects

This PR refactors the skwasm renderer to use DisplayList objects as its main model objects instead of using Skia objects directly. Then, at render time, we dispatch the display list commands to the skia surface. This is a preparatory step for impeller on web.

* Some build rules were reworked in order to allow DisplayList to compile via emscripten
* SkPath and SkImage are still used as the main model objects in skwasm. As of right now, DisplayList just thinly wraps these objects, so this is the minimal possible change for now. I will have to refactor this somewhat further when preparing for actual impeller adoption.
* Several special cased code paths in skwasm were removed, as they are taken care of by DisplayList itself. This includes shadow drawing, determining when to enable dithering, and determining the right clamp value for filters.

eyebrowsoffire commented on engine/src/flutter/display_list/display_list_unittests.cc:
```diff
@@ -1224,7 +1230,8 @@ TEST_F(DisplayListTest, SingleOpsMightSupportGroupOpacityBlendMode) {
     RUN_TESTS2(canvas.DrawDisplayList(display_list);, false);
   }
-  RUN_TESTS2(canvas.DrawTextBlob(GetTestTextBlob(1), 0, 0, paint);, false);
+  RUN_TESTS2(canvas.DrawText(DlTextSkia::Make(GetTestTextBlob(1)), 0, 0, paint);
```
I think there are already test cases for text frames lower down in this file.

flar reviewed (commented):
Deferring to @harryterkelsen now...

flar reviewed (approved):
Can't "unrequest changes" - must approve.

For the record, I didn't really review any of the code that's left, assuming that @harryterkelsen took care of that...

ksokolovskyi commented on engine/src/flutter/lib/web_ui/lib/src/engine/skwasm/skwasm_impl/paint.dart:
```diff
@@ -26,16 +24,21 @@ class SkwasmPaint implements ui.Paint {
     final shaderHandle = _shader?.handle;
     if (shaderHandle != null) {
       paintSetShader(rawPaint, shaderHandle);
-      if (_shader!.isGradient) {
-        paintSetDither(rawPaint, true);
-      }
```
Hi @eyebrowsoffire, sorry for disturbing you, but I am wondering why gradient dithering was removed?

eyebrowsoffire replied:
> Hi @eyebrowsoffire, sorry for disturbing you, but I am wondering why gradient dithering was removed?

Good question. This special casing (setting the dither flag when using a gradient shader) is implicitly handled by the `DisplayList` machinery lower down the stack. So we no longer need this code path. Functionally it ends up the same in the end.

ksokolovskyi replied:
Oh, got it, thanks a lot for the explanation!
